Terraform Docs - HCP TF SSO Linking (#2230)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/users-teams-organizations/single-sign-on/linking-user-account.mdx

<!-- END: TFC:only name:hcp-eu -->

## Add SSO Identity Link
The first time you sign in to an organization using single sign-on (SSO), HCP Terraform links your SSO identity to your user account. After HCP Terraform creates this link, you must use that account to access the organization through SSO. You may link your SSO identity from an active login session or from the organization SSO sign-in page.

### Linking SSO identity from an active login session
To link your SSO identity to your HCP Terraform user account from from an active login session you must be invited or active member of the organization. Sign into HCP Terraform with an account that does not have an SSO identity.  Next, select the SSO organization from the **Choose an organization** drop-down in the menu bar.  Click the **Authorize** button, then complete SSO authentication.

Once you have successfully authenticated with your SSO provider, HCP Terraform links your SSO identity to your HCP Terraform account.

### Linking SSO identity from the Organization SSO sign-in page

> To link your SSO identity to your HCP Terraform user account from the [Organization sign-in page](https://app.terraform.io/login/sso), if your SSO identity is not yet linked:

Reviewer: Should we possibly remove if your SSO identity is not yet linked? 

To link an SSO identity that isn’t linked yet to your HCP Terraform user account from the [Organization sign-in page](https://app.terraform.io/login/sso) :

Author: hey I'm a little confused by your feedback why do we need to remove "if your SSO identity is not yet linked"?

Reviewer: hey! Yeah, so sorry. Like, at the very end you added that small line and I suggested to remove it. I figured we can try to restructure the sentence to where that condition at the end, "if your sso identity is not yet linked" can be more, roped into the actual sentence itself. To where it becomes:

"To link an SSO identity that isn't yet linked to your HCP Terraform user account from the organization sign in page:"

1. Enter your organization name and click the **Continue** button.

> 2. Authenticate with your identity provider (IdP).  After you have authenticated with your IdP, HCP Terraform displays the **Link SSO to HCP Terraform** step.

Reviewer: Maybe we can figure out how to try to state the action which is first then talk about the result in the next step. 

1. Enter your organization name and click **continue**.
2. Sign in with your identity provider (IdP).
3. After you sign in, HCP Terraform displays the **Link SSO to HCP Terraform** step. Enter the email address or username of the HCP Terraform account you want to link to your SSO Identity.

Reviewer: Maybe instead:

After signing in, HCP Terraform...

Author: > Maybe we can figure out how to try to state the action which is first then talk about the result in the next step.

I think each step should start with an action to be taken by the reader. It makes it a clearer what they must do

Reviewer: Perfect! This is good for me to know with doc writing 😁

3. Enter the email address or username of the HCP Terraform account you want to link to your SSO identity.

> After you have entered the email address or username of your HCP Terraform account, HCP Terraform will prompt you based on the following conditions:

Reviewer: Totally all suggestions you see are optional - just to see if we can build on it. But this is already great!! 

After you enter the email address or username of your HCP Terraform account, HCP Terraform prompts you based on these scenarios: 

I wonder though in documentation, do we want like verbiage based on user being on _actively_ on that step?

Like example, enter, prompts, versus have entered, will prompt?

Author: @mattschouten should the tense be active or past?

Reviewer 2: The style guide calls for active voice, simple present tense (see https://github.com/hashicorp/web-unified-docs/blob/main/docs/style-guide/top-12.md)

Reviewer 2: So in this case, starting with "After you have entered" is fine, because it shows sequence.  It'd be better to write "..., HCP Terraform prompts you..." (vs "will prompt")

- If the email address or username belongs to an account that is already linked to a HashiCorp Cloud Platform (HCP) account, HCP Terraform prompts you to sign in through HCP to complete the link.  Refer to [Linked HCP and HCP Terraform Accounts](/terraform/cloud-docs/users-teams-organizations/users#linked-hcp-and-hcp-terraform-accounts) for more details.
- If the email address or username belongs to an existing HCP Terraform account that is not linked to an HCP account, HCP Terraform prompts you to enter your HCP Terraform account password to complete the link to your SSO identity.

> - If the email address or username is not associated with any HCP Terraform account, HCP Terraform prompts you to create a new account.  After you have created a new account, HCP Terraform links that account to your SSO identity.

Reviewer: Optional as well - 

If the email address or username is not associated with any HCP Terraform account, HCP Terraform prompts you to create a new account and then links it to your SSO identity.

#### What happens next

> If SSO linking succeeds, HCP Terraform signs you in and grants access to the organization and permanently associates your SSO identity with your account for that organization. Future SSO sign-ins use this SSO Identity to authorize signing into your HCP Terraform account.

Reviewer: Hmm, I wonder if there's a way we can split it up. Like, the order of events. Something like-

If SSO linking succeeds, HCP Terraform signs you in, grants you access to the  organization, and permanently associates your SSO identity with your account for that organization. 

And then like, have failures afetr based off account type and break it down into two possible paths.

If SSO linking fails when creating a new account... 
    1. Accounts not linked to an HCP account: ... 
    2. Accounts linked to HCP account: ...

Author: I'll see if i can work this into sections 👍

<!-- END: TFC:only name:hcp-eu -->

## Add SSO Identity Link
The first time you sign in to an organization using single sign-on (SSO), HCP Terraform links your SSO identity to your user account. After HCP Terraform creates this link, you must use that account to access the organization through SSO. You may link your SSO identity from an active login session or from the organization SSO sign-in page.

### Linking SSO identity from an active login session

> To link your SSO identity to your HCP Terraform user account from from an active login session you must be invited or active member of the organization. Sign into HCP Terraform with an account that does not have an SSO identity.  Next, select the SSO organization from the **Choose an organization** drop-down in the menu bar.  Click the **Authorize** button, then complete SSO authentication.

Reviewer 3: ```suggestion
To link your SSO identity to your HCP Terraform user account from an active login session you must be invited or an active member of the organization. First, sign into HCP Terraform with an account that does not have an SSO identity.  Next, select the SSO organization from the **Choose an organization** drop-down in the menu bar.  Click the **Authorize** button, then complete SSO authentication.
```

Reviewer 3: Added "First", since the first instruction isn't the first sentence of the paragraph, it seems a bit clearer to indicate "time to do stuff now" to the reader.

> To link your SSO identity to your HCP Terraform user account from an active login session you must be invited or an active member of the organization. First, sign into HCP Terraform with an account that does not have an SSO identity.  Next, select the SSO organization from the **Choose an organization** drop-down in the menu bar.  Click the **Authorize** button, then complete SSO authentication.

Reviewer 2: ```suggestion
To link your SSO identity to your HCP Terraform user account from an active login session, you must be invited to or an active member of the organization in HCP Terraform.  When you have been invited or added to the organization, sign into HCP Terraform with an account that does not have an SSO identity.  Next, select the SSO organization from the **Choose an organization** drop-down in the menu bar.  Click the **Authorize** button, then complete SSO authentication.
```

Please check the correctness of my edits.  Nice work adding the requirement step.

3. Enter the email address or username of the HCP Terraform account you want to link to your SSO identity.

After you have entered the email address or username of your HCP Terraform account, HCP Terraform will prompt you based on the following conditions:

- If the email address or username belongs to an account that is already linked to a HashiCorp Cloud Platform (HCP) account, HCP Terraform prompts you to sign in through HCP to complete the link.  Refer to [Linked HCP and HCP Terraform Accounts](/terraform/cloud-docs/users-teams-organizations/users#linked-hcp-and-hcp-terraform-accounts) for more details.
- If the email address or username belongs to an existing HCP Terraform account that is not linked to an HCP account, HCP Terraform prompts you to enter your HCP Terraform account password to complete the link to your SSO identity.
- If the email address or username is not associated with any HCP Terraform account, HCP Terraform prompts you to create a new account.  After you have created a new account, HCP Terraform links that account to your SSO identity.

#### What happens next
**If SSO linking succeeds**

> You’re signed in, granted access to the organization, and your SSO identity is permanently linked to your account. Future SSO sign-ins use this identity automatically.

Reviewer 2: ```suggestion
**If SSO linking succeeds**, you’re signed in, granted access to the organization, and your SSO identity is permanently linked to your account. Future SSO sign-ins use this identity automatically.
```

> **If SSO linking fails**

Reviewer 2: Could this whole section be simplified while remaining useful to read something like

`**If SSO linking fails**, HCP Terraform will display an error message.  Your HCP Terraform account will not have access to the organization.`

I'm thinking the details of whether you're signed in or not might not matter much to the user, because they're in the context of trying to link identities.  My suggested rewrite focuses on that.

The outcome depends on your account:
1. Not linked to an HCP account
  - Returned to the linking form with an error
  - Not signed in
2. Linked to an HCP account
  - Signed in, but shown an error
  - No access to the organization

### Important notes

> - You can link an SSO identity to only one account per organization.

Reviewer 2: These two lines are related and too important to be this far down in the document.  I'd move the first line to a new paragraph in the intro (between "You can add and remove SSO identity links for all providers..." and "@include 'eu/sso.mdx'"), and leave the second implicit.  So the new paragraph, way higher in the doc, would read:

You may link an SSO identity to only one account per organization.

Author: Good callout. I think this section actually covers that first point already:

>  Link a user account for single sign-on
>
> You have an SSO identity for every SSO-enabled HCP Terraform 
> organization. HCP Terraform links each SSO identity to a single HCP 
> Terraform user account. This link determines which account you can use to access each organization.
>
> You can add and remove SSO identity links for all providers, including Microsoft Entra ID, Okta, and SAML.

"HCP Terraform links each SSO identity to a single HCP Terraform user account." implicitly says you can link an SSO Identity to only one account

Reviewer 2: You're right!

## Add SSO Identity Link

> The first time you sign in to an organization using single sign-on (SSO), HCP Terraform links your SSO identity to a user account. After HCP Terraform creates this link, you must use that account to access the organization through SSO. SSO linking behavior depends on how you start the sign-in process:

Reviewer 2: ```suggestion
The first time you sign in to an organization using single sign-on (SSO), HCP Terraform links your SSO identity to your user account. After HCP Terraform creates this link, you must use that account to access the organization through SSO. You may link your SSO identity from an active login session or from the organization SSO sign-in page.
```

> ### From an active session

Reviewer 2: ```suggestion
### Linking SSO identity from an active login session
```

A bit wordy, but more meaningful if the above heading isn't visible

> If you are already signed in to HCP Terraform with an account that does not have an SSO identity, and you initiate SSO, HCP Terraform links your SSO identity to that account after successful SSO authentication.

Reviewer 2: ```suggestion
To link your SSO identity to your HCP Terraform user account from from an active login session, sign into HCP Terraform with an account that does not have an SSO identity.  Next, select the SSO organization from the **Choose an organization** drop-down in the menu bar.  Click the **Authorize** button, then complete SSO authentication.

Once you have successfully authenticated with your SSO provider, HCP Terraform links your SSO identity to your HCP Terraform account.
```
Provided more instructions in a more directive way.

Please double-check that the instructions are correct—I used the videos as the source.

Author: So there's a small step that's not captured in the demo: The user must be invited to the Organization to see if from the Org list page. In most cases, customers onboard their SSO user/SSO Identity link by using the SSO Org sign-in flow. 

🤔 Lemme stew on this a bit and I'll propose something else here

Author: took a stab, let me know your thoughts

> ### From the Organization SSO sign-in page

Reviewer 2: ```suggestion
### Linking SSO identity from the Organization SSO sign-in page
```

> If you start from the [Organization sign-in page](https://app.terraform.io/login/sso) and your SSO identity is not yet linked:

Reviewer 2: ```suggestion
To link your SSO identity to your HCP Terraform user account from the [Organization sign-in page](https://app.terraform.io/login/sso), if your SSO identity is not yet linked:
```

> 1. After you authenticate with your identity provider (IdP), HCP Terraform displays the **Select Account** step.

Reviewer 2: ```suggestion
1. Enter your organization name and click the **Continue** button.
1. Authenticate with your identity provider (IdP).  After you have authenticated with your IdP, HCP Terraform displays the **Link SSO to HCP Terraform** step.
```
From the video, the heading on the page is "Link SSO to HCP Terraform", but please correct me if I'm wrong.

> 2. Enter the email address or username of the account you want to link.

Reviewer 2: ```suggestion
1. Enter the email address or username of the HCP Terraform account you want to link to your SSO identity.
```

> Based on what you enter, HCP Terraform follows one of the following paths:

Reviewer 2: ```suggestion
After you have entered the email address or username of your HCP Terraform account, HCP Terraform will prompt you based on the following conditions:
```
Showing the sequence a bit more.  I don't quite like the part after the comma.  Open to other/better ideas.

> - HCP Linked account: If the email address or username belongs to an HCP Linked account, HCP Terraform prompts you to sign in through HCP to complete the link. Refer to [Linked HCP and HCP Terraform Accounts](/terraform/cloud-docs/users-teams-organizations/users#linked-hcp-and-hcp-terraform-accounts) for more details.

Reviewer 2: ```suggestion
- If the email address or username belongs to an account that is already linked to a HashiCorp Cloud Platform (HCP) account, HCP Terraform prompts you to sign in through HCP to complete the link.  Refer to [Linked HCP and HCP Terraform Accounts](/terraform/cloud-docs/users-teams-organizations/users#linked-hcp-and-hcp-terraform-accounts) for more details.
```

> - Existing HCP Terraform account: If the account exists but is a regular HCP Terraform account, HCP Terraform prompts you to enter your password to link the account.

Reviewer 2: ```suggestion
- If the email address or username belongs to an existing HCP Terraform account that is not linked to an HCP account, HCP Terraform prompts you to enter your HCP Terraform account password to complete the link to your SSO identity.
```

Author: 🙌 thank you this is a much better wording. Really appreciate your expertise here

> - No existing account: If the email address or username is not associated with any account, HCP Terraform prompts you to create a new account, then links that account to your SSO identity.

Reviewer 2: ```suggestion
- If the email address or username is not associated with any HCP Terraform account, HCP Terraform prompts you to create a new account.  After you have created a new account, HCP Terraform links that account to your SSO identity.
```

> #### What happens next

Reviewer 2: Recommend turning this section into a non-bulleted paragraph.

After linking succeeds:
- HCP Terraform permanently associates your SSO identity with your account for that organization.
- HCP Terraform signs you in and grants access to the organization
- Future SSO sign-ins use this SSO Identity to authorize signing into your HCP Terraform account.

> ### Important notes

Reviewer 2: I don't think this section is overkill—I don't think the info in this sentence exists anywhere else in the docs.  Consider moving the first item ("only one account") to higher on the page, maybe in the intro section.

The second and third items could each become their own paragraph.  For the second, it'd be worth covering what happens if the link fails (errors, steps to take, etc).  For the third, I'm curious about "may"—probably worth getting crisp about what happens when linking fails, and what "does not grant access to the organization" means to the user.

Author: btw I moved the bullet point about how failures are handled into the "What happens" next section. What I wrote is rather wordy but follows the established phrasing

- You can link an SSO identity to only one account per organization.
- If you try to link a different SSO identity to an already-linked account, the link fails.

> - If linking fails, HCP Terraform may still sign you in, but it does not grant access to the organization.

Author: this section may be overkill
